### PR TITLE
Sign with entitlements when required, including app bundles

### PR DESCRIPTION
**Pull request by bhearsum**
Previous version of this was #315, which didn't sign app bundles with entitlements ever (only inner files), which caused a Firefox start-up crash. I ended up doing a bit of refactoring to make this cleaner (at least, IMO).

Tested Mozilla VPN & Firefox signing. I can send signed binaries privately if you want to inspect them.

### CODE REVIEW

File: iscript/src/iscript/mac.py
Comment (escapewindow) [inline iscript/src/iscript/mac.py]: Is there a reason we have both `hardened_runtime_files` and `no_entitlements_files`? (Will we need some files with no `-o runtime` at some point?)

Similarly, since these blocks are mutually exclusive, is there a reason why `wg` and `wireguard-go` are in both `hardened_runtime_files` and `no_entitlements_files` ?

If we see needing more complex special configs per file, we may want a dictionary keyed by filename. We don't have to go down that route unless we see it coming shortly. I mainly want to make sure we want two config settings here, when it seems like one would solve the issue at hand.

File: iscript/src/iscript/mac.py
Comment (escapewindow) [inline iscript/src/iscript/mac.py]: Nit: I tend to prefer explicitly setting kwargs, e.g. `file_=file_`. Non-blocking if it passes the linter.

File: iscript/src/iscript/mac.py
Comment (escapewindow) [inline iscript/src/iscript/mac.py:300]: Will we ever need to not use an entitlements file or hardened runtime on the top level app?

**Review by escapewindow — CHANGES_REQUESTED** [review]
It's possible we can land this without changes, depending on the answers to my questions.

File: iscript/src/iscript/mac.py
Comment (bhearsum) [reply iscript/src/iscript/mac.py]: I can try to simplify this, I agree it's not great, and it ended up this why through evolution, not design...

Truthfully though, I'm treating this is a stop-gap solution until we have a proper manifest for specifying these things for an entire app (ideally one that comes in as part of of the task payload).

Comment (escapewindow) [reply iscript/src/iscript/mac.py]: LMK. If it turns out this is the best option for reasons, we can land as is. If there's a quick way to improve it, let's do that.

File: iscript/src/iscript/mac.py
Comment (bhearsum) [reply iscript/src/iscript/mac.py:300]: I don't know! I suspect not. I didn't think a YAGNI philosophy would make sense here though, since it was trivial to support all paths.

File: iscript/src/iscript/mac.py
Comment (bhearsum) [reply iscript/src/iscript/mac.py]: Okay, so after some review I believe the following is true:
* We have to keep `no_entitlements_files` (or something equivalent), because we cannot sign two of the vpn binaries with entitlements.
* If we want to preserve the behaviour of _not_ signing with a hardened runtime by default, we have to keep `hardened_runtime_files` (to enable it for these same files without enabling entitlements). The alternative is to always include `-o runtime` in the signing command, which may or may not break something.

Do you have a preference here?

Comment (escapewindow) [reply iscript/src/iscript/mac.py]: > Okay, so after some review I believe the following is true:
> 
>     * We have to keep `no_entitlements_files` (or something equivalent), because we cannot sign two of the vpn binaries with entitlements.

The `no_entitlements_files` check is in an `elif` where we've already hit matched the `if` block (`hardened_runtime_files`). As far as I can see, nothing uses `no_entitlements_files` without `hardened_runtime_files`. Do you think we'll still need both in some future use case?

**Review by escapewindow — APPROVED** [review]
I'm going to mark this approved, to unblock you when you think you've got this right.

File: iscript/src/iscript/mac.py
Comment (escapewindow) [reply iscript/src/iscript/mac.py]: (Marked approved. If the answer is `no`, we can remove `no_entitlements_files`. If the answer is `yes`, we can land as is.)

Comment (bhearsum) [reply iscript/src/iscript/mac.py]: Oh, I see. Yes, we could get rid of `no_entitlements_files`, but I'd probably rename `hardened_runtime_files` to `hardened_runtime_only_files` for clarity (otherwise one might expect them to still get entitlements, and anything else _not_ to have hardened runtime).

**Comment by bhearsum** [comment]
@escapewindow - do you want another look at this before it lands?

File: iscript/src/iscript/mac.py
Comment (escapewindow) [inline iscript/src/iscript/mac.py]: Does this mean we no longer sign libclearkey with entitlements?

**Review by escapewindow — COMMENTED** [review]
Other than the libclearkey question, this looks good! Thank you!

File: iscript/src/iscript/mac.py
Comment (escapewindow) [reply iscript/src/iscript/mac.py]: (Did we check to make sure Firefox still starts up with this patch applied?)

Comment (bhearsum) [reply iscript/src/iscript/mac.py]: I did! But I'm going to do some log comparison to make sure signing command invocations haven't changed vs. current deployed code.

Comment (bhearsum) [reply iscript/src/iscript/mac.py]: Indeed! Thanks for catching this. Firefox still runs with it, but I imagine this would mean Widevine would break.

**Comment by bhearsum** [comment]
With the latest patch, codesign commands are 100% the same as the current ones for Firefox.

**Pull request merged**
